### PR TITLE
11 feature 11 write account entity class

### PR DESCRIPTION
**Pull request by A-Bunch-Of-Atoms**
Completed skeleton Account entity class. Updated branch to match new main branch.

Account class will need to be updated when Location, Gender, and Filters are completed.

### CODE REVIEW

File: app/src/main/java/com/courseproject/tindar/entity/Account.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.courseproject.tindar.entity;
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entity/Account.java:1]: Can we have a folder name entities as plural since we will put multiple entities in this folder?

File: app/src/main/java/com/courseproject/tindar/entity/Account.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.util.ArrayList;
+
+class Account {
+    int status;
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entity/Account.java:6]: Pls add comment to indicate what status each integer means. Alternatively, you can have isActive boolean if active / suspended are only two status we will have.

File: app/src/main/java/com/courseproject/tindar/entity/Account.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    String id;
+    String email;
+    String password;
+    String displayName;
+    String firstName;
+    String lastName;
+    String aboutMe;
+    String profilePictureLink;
+    String birthdate;
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entity/Account.java:15]: Can we have a type Date?

File: src/main/java/account/Account.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package account;
```
Comment (jenniferjinyoungyang) [inline src/main/java/account/Account.java:1]: Can we delete src/main/java/account folder?

File: app/src/main/java/com/courseproject/tindar/entity/Account.java (hunk 1 above)
Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/entity/Account.java:1]: done

File: app/src/main/java/com/courseproject/tindar/entity/Account.java (hunk 2 above)
Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/entity/Account.java:6]: fixed, status is now boolean

File: src/main/java/account/Account.java (hunk 4 above)
Comment (A-Bunch-Of-Atoms) [reply src/main/java/account/Account.java]: was removed on my end, for some reason wasn't deleted on github. that has been done.

File: app/src/main/java/com/courseproject/tindar/entity/Account.java (hunk 3 above)
Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/entity/Account.java]: changed.

File: app/src/main/java/com/courseproject/tindar/entity/Account.java (hunk 2 above)
Comment (jenniferjinyoungyang) [reply app/src/main/java/com/courseproject/tindar/entity/Account.java]: Can we name attribute isActiveStatus instead of status? So it is more clear.

Comment (A-Bunch-Of-Atoms) [reply app/src/main/java/com/courseproject/tindar/entity/Account.java]: Done.

**Review by jenniferjinyoungyang — APPROVED** [review]

**Review by kresimirgotovac — APPROVED** [review]

**Review by navarhontes — CHANGES_REQUESTED** [review]
1. Pretty much everything here should be private 
2. Getters and setters since we're making things private 
3. We should make a habit of actually writing tests 
4. I have a strong feeling that we're going to end up changing our implementation of Profile stuff. So you might want to have methods about about me, display name, and profile picture as an interface that Account implements.

File: app/src/main/java/com/courseproject/tindar/entities/AccountController.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
+package com.courseproject.tindar.entities;
+
+import android.location.Location;
+
+import java.util.ArrayList;
+import java.util.Date;
+
+public interface AccountController {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/AccountController.java:8]: A controller is something that interacts with frontend, and is not used as an interface for the entity class. Please refer to the UserLoginCleanArchitecture repo. You can maybe use factory design similarly with that repo?

File: app/src/main/java/com/courseproject/tindar/entities/AccountController.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    /** Creates an account after receiving an email,
+     * password, display name, first name, and last name.
+     *
+     * @param email the email to be associated with the account
+     * @param password the password to be associated with the account
+     * @param display the display name to be associated with the account
+     * @param first the first name to be associated with the account
+     * @param last the last name to be associated with the account
+     */
+    default void createAccount(String email, String password,
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/AccountController.java:18]: why default here? Why are we implementing all the methods in the interface?

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+    private String lastName;
+    private String aboutMe;
+    private String profilePictureLink;
+    private ArrayList<String> matchList;
+    private ArrayList<String> likeList;
+    private ArrayList<String> blockList;
+    private ArrayList<String> recentlyViewed;
+    private Date birthdate;
+    private Location location;
+    private Gender gender;
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:25]: Can we have Location and Gender as String to make it simpler? The frontend will only give a few options to select from so frontend will control what values a user can enter for those two attributes.

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            return "Active";
+        }
+        else {
+            return "Suspended";
+        }
+    }
+    /** Flips the account status from Active to Suspendedor vice versa.
+     *
+     */
+    void setAccountStatus() {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:73]: I believe it would be more clear to pass the argument...like setAccountStatus(boolean isActive). The current implementation looks more like toggle than setting.

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+     * @return the display name associated with the account
+     */
+    String getDisplayName() {
+        return this.displayName;
+    }
+    /** Changes the display name associated with the account.
+     *
+     * @param display the new display name associated with the account
+     */
+    void setDisplayName(String display) {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:122]: Can we have parameter name as displayName instead of display to make it clearer?

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+     * @return the first name associated with the account
+     */
+    String getFirstName() {
+        return this.firstName;
+    }
+    /** Changes the first name associated with the account.
+     *
+     * @param first the new first name associated with the account
+     */
+    void setFirstName(String first) {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:136]: firstName instead of first?

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+     * @return the last name associated with the account
+     */
+    String getLastName() {
+        return this.lastName;
+    }
+    /** Changes the last name associated with the account.
+     *
+     * @param last the new last name associated with the account
+     */
+    void setLastName(String last) {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:150]: lastName instead of last?

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+     * @return the about me section associated with the account
+     */
+    String getAboutMe() {
+        return this.aboutMe;
+    }
+    /** Changes the about me section of the account.
+     *
+     * @param about the new about me section of the account
+     */
+    void setAboutMe(String about) {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:164]: aboutMe instead of about?

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+     * @return the link to the profile image associated with the account
+     */
+    String getProfilePictureLink() {
+        return this.profilePictureLink;
+    }
+    /** Changes the profile picture associated with the account.
+     *
+     * @param picture the link to the new profile picture associated with the account
+     */
+    void setProfilePictureLink(String picture) {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:178]: profilePictureLink instead of picture?

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+     * @return the birthdate associated with the account
+     */
+    Date getBirthdate() {
+        return this.birthdate;
+    }
+    /** Changes the birthdate associated with the account.
+     *
+     * @param date the new birthdate associated with the account
+     */
+    void setEmail(Date date) {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:277]: should be setBirthday?

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+        this.location = location;
+    }
+    /** Returns a user's gender from their account in the form of a Gender.
+     *
+     * @return the gender associated with the account
+     */
+    Gender getGender() {
+        return this.gender;
+    }
+    // TODO: Create setGender(). Requires Gender class to be created.
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:301]: we can create setGender if we make gender String

**Review by jenniferjinyoungyang — DISMISSED** [review]

**Review by kresimirgotovac — DISMISSED** [review]

**Comment by navarhontes** [comment]
Thanks, Caullin. Just one more change that I will request before approving this: toggle account status rather than set account status seems like bad design, since that means people using it have to check account status, use an if block, then call it or not call it. Instead, setAccountStatus should probably do nothing but return False if the inputted string is not acceptable (i.e. an approved option). 

Some stuff that we don't have to handle right now but that I want to note:
1. I have a pretty strong feeling that the list stuff and profile information will need to be reimplemented at some other point, but we don't have to focus on that right now. I just want to give you a heads up since that's likely stuff that will get rewritten. 
2. I'll approve this without tests since some stuff will probably need to be reimplemented and the code so far is simple getting and setting, but note also that this sets a bad precedent.

**Comment by navarhontes** [comment]
Like I said before, we should either handle or entirely prevent the case where a status is invalid. One option I had proposed earlier: 
> Instead, setAccountStatus should probably do nothing but return False if the inputted string is not acceptable (i.e. an approved option); and update status and return True if it is a valid status.

Thanks again, Caullin.

**Review by jenniferjinyoungyang — DISMISSED** [review]

**Comment by jenniferjinyoungyang** [comment]
@navarhontes @A-Bunch-Of-Atoms 
I just noticed setAccountStatus takes an argument boolean. So we are not inputting String. If you want to make it more clear, you can maybe do, 

`public void setIsActiveStatus(boolean isActiveStatus) { this.isActiveStatus = isActiveStatus}`

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+        this.birthdate = null;
+        this.location = null;
+        this.gender = null;
+        this.filters = null;
+    }
+    /** Returns the status of an account in the form of a String.
+     *
+     * @return the status of the account
+     */
+    String getAccountStatus() {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:61]: I just noticed we don't have access modifiers for the methods here. Don't we want to have it public?

**Comment by navarhontes** [comment]
Thanks for pointing out that setAccountStatus takes a boolean, Jennifer. I'm currently debating over how likely it is that we will need other statuses and whether it would be helpful or not to set ourselves up to easily support additional statuses. The most likely statuses I can think of are "banned" and "suspended by admin," but I don't know that we'll have time to implement that anyway. So you can write this up as an issue if you'd like but I'll approve this pull request in the meantime.

**Review by navarhontes — DISMISSED** [review]

**Review by jenniferjinyoungyang — APPROVED** [review]

**Review by yiwxng — APPROVED** [review]

**Comment by navarhontes** [comment]
Quick thing: why was this necessary? 
> @SuppressWarnings("FieldMayBeFinal")

File: app/src/main/java/com/courseproject/tindar/entities/Account.java [hunk 17]
```diff
@@ -0,0 +1,317 @@
+package com.courseproject.tindar.entities;
+
+import java.util.ArrayList;
+import java.util.Date;
+
+@SuppressWarnings("FieldMayBeFinal")
```
Comment (navarhontes) [inline app/src/main/java/com/courseproject/tindar/entities/Account.java:6]: Quick thing: why was this necessary?

**Review by navarhontes — APPROVED** [review]

File: app/src/main/java/com/courseproject/tindar/entities/Account.java (hunk 17 above)
Comment (navarhontes) [reply app/src/main/java/com/courseproject/tindar/entities/Account.java:6]: Caullin's response: 
> I kept getting warnings about the arraylists saying "field may be final"
> And it was suggesting me to make them final
> Kinda annoying when i was looking through the errors warnings for problems
